File: apps/ios/PhrenKit/Sources/PhrenKit/Metadata/JSRegex.swift
import Foundation

/// Thin wrapper over NSRegularExpression with JavaScript-like semantics, so the
/// patterns transcribed from packages/cli/src/content/metadata.ts behave
/// identically to their TypeScript originals.
struct JSRegex: @unchecked Sendable {
    let regex: NSRegularExpression
    // NSCache is thread-safe and evictable; patterns/options are the complete key.
    private static let compiled: NSCache<NSString, NSRegularExpression> = {
        let cache = NSCache<NSString, NSRegularExpression>()
        cache.countLimit = 256
        return cache
    }()

    private static func expression(_ pattern: String, options: NSRegularExpression.Options) -> NSRegularExpression {
        let key = "\(options.rawValue):\(pattern)" as NSString
        if let cached = compiled.object(forKey: key) { return cached }
        let regex = try! NSRegularExpression(pattern: pattern, options: options)
        compiled.setObject(regex, forKey: key)
        return regex
    }

    init(_ pattern: String, caseInsensitive: Bool = false) {
        var options: NSRegularExpression.Options = []
        if caseInsensitive { options.insert(.caseInsensitive) }
        // Patterns are compile-time constants transcribed from the CLI; a failure
        // to parse is a programmer error.
        // swiftlint:disable:next force_try
        self.regex = Self.expression(pattern, options: options)
    }

    init(multiline pattern: String) {
        // JS `m` flag: ^/$ match at line boundaries.
        // swiftlint:disable:next force_try
        self.regex = Self.expression(pattern, options: [.anchorsMatchLines])
    }

    func test(_ s: String) -> Bool {
        firstMatch(in: s) != nil
    }

    func firstMatch(in s: String) -> NSTextCheckingResult? {
        regex.firstMatch(in: s, range: NSRange(s.startIndex..., in: s))
    }

    /// Equivalent of `s.match(re)?.[group]` for the first match.
    func group(_ s: String, _ group: Int = 1) -> String? {
        guard let m = firstMatch(in: s) else { return nil }
        return Self.substring(s, m, group)
    }

    /// All capture-group-1 values, like `[...s.matchAll(re)].map(m => m[1])`.
    func allGroups(_ s: String, _ group: Int = 1) -> [String] {
        regex.matches(in: s, range: NSRange(s.startIndex..., in: s))
            .compactMap { Self.substring(s, $0, group) }
    }

    /// All whole-match values, like `s.match(/.../g)`.
    func allMatches(_ s: String) -> [String] {
        regex.matches(in: s, range: NSRange(s.startIndex..., in: s))
            .compactMap { Self.substring(s, $0, 0) }
    }

    /// Equivalent of `s.replace(re, replacement)` with a global regex. The
    /// replacement string is literal (no $-templates).
    func replaceAll(_ s: String, with replacement: String) -> String {
        regex.stringByReplacingMatches(
            in: s,
            range: NSRange(s.startIndex..., in: s),
            withTemplate: NSRegularExpression.escapedTemplate(for: replacement)
        )
    }

    /// Equivalent of a non-global `s.replace(re, replacement)`: first match only.
    func replaceFirst(_ s: String, with replacement: String) -> String {
        guard let m = firstMatch(in: s), let range = Range(m.range, in: s) else { return s }
        return s.replacingCharacters(in: range, with: replacement)
    }

    static func substring(_ s: String, _ m: NSTextCheckingResult, _ group: Int) -> String? {
        guard group < m.numberOfRanges, let range = Range(m.range(at: group), in: s) else { return nil }
        return String(s[range])
    }
}

extension String {
    /// JS `String.prototype.trim()` trims the same whitespace set as
    /// `\s`, which `trimmingCharacters(in: .whitespacesAndNewlines)` matches
    /// closely enough for markdown content.
    var jsTrimmed: String {
        trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// JS `s.replace(/\s+/g, " ")`.
    var collapsedWhitespace: String {
        JSRegex(#"\s+"#).replaceAll(self, with: " ")
    }
}

File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/JSRegex.kt
package com.phren.kit

import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * java.util.regex wrapper with JavaScript semantics, so the patterns
 * transcribed from packages/cli/src/content/metadata.ts behave identically to
 * their TypeScript originals (port of JSRegex.swift).
 *
 * The one real gap between the engines is `\s`: JS matches Unicode whitespace
 * (NBSP, U+2028, BOM, ...) while Java's `\s` is ASCII-only. [jsPattern]
 * rewrites every `\s` to the JS set; `\w`/`\d` are ASCII in both.
 */
class JSRegex private constructor(val pattern: Pattern) {
    constructor(source: String, caseInsensitive: Boolean = false) :
        this(Pattern.compile(jsPattern(source), if (caseInsensitive) Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE else 0))

    fun test(s: String): Boolean = pattern.matcher(s).find()

    fun firstMatch(s: String): Matcher? {
        val m = pattern.matcher(s)
        return if (m.find()) m else null
    }

    /** Equivalent of `s.match(re)?.[group]` for the first match. */
    fun group(s: String, group: Int = 1): String? = firstMatch(s)?.let { substring(it, group) }

    /** All capture-group values, like `[...s.matchAll(re)].map(m => m[1])`. */
    fun allGroups(s: String, group: Int = 1): List<String> {
        val m = pattern.matcher(s)
        val out = mutableListOf<String>()
        while (m.find()) substring(m, group)?.let(out::add)
        return out
    }

    /** All whole-match values, like `s.match(/.../g)`. */
    fun allMatches(s: String): List<String> = allGroups(s, 0)

    /** `s.replace(re, replacement)` with a global regex; replacement is literal. */
    fun replaceAll(s: String, replacement: String): String =
        pattern.matcher(s).replaceAll(Matcher.quoteReplacement(replacement))

    /** Non-global `s.replace(re, replacement)`: first match only. */
    fun replaceFirst(s: String, replacement: String): String =
        pattern.matcher(s).replaceFirst(Matcher.quoteReplacement(replacement))

    companion object {
        /** JS `^`/`$` at line boundaries (the `m` flag). */
        fun multiline(source: String): JSRegex = JSRegex(Pattern.compile(jsPattern(source), Pattern.MULTILINE))

        fun substring(m: Matcher, group: Int): String? =
            if (group <= m.groupCount()) m.group(group) else null

        /** The members of JS's `\s` (ECMA-262 WhiteSpace + LineTerminator). */
        private const val JS_WS = "\\t\\n\\x{0B}\\f\\r \\x{A0}\\x{1680}\\x{2000}-\\x{200A}\\x{2028}\\x{2029}\\x{202F}\\x{205F}\\x{3000}\\x{FEFF}"

        /** Rewrite `\s` / `\S` to the JS whitespace set, inside or outside a class. */
        internal fun jsPattern(source: String): String {
            val out = StringBuilder()
            var inClass = false
            var i = 0
            while (i < source.length) {
                val c = source[i]
                if (c == '\\' && i + 1 < source.length) {
                    val n = source[i + 1]
                    when {
                        n == 's' -> out.append(if (inClass) JS_WS else "[$JS_WS]")
                        n == 'S' && !inClass -> out.append("[^$JS_WS]")
                        else -> out.append(c).append(n)
                    }
                    i += 2
                    continue
                }
                if (c == '[' && !inClass) inClass = true
                else if (c == ']' && inClass) inClass = false
                out.append(c)
                i++
            }
            return out.toString()
        }

        private val JS_WS_CHARS: (Char) -> Boolean = { ch ->
            ch == '\t' || ch == '\n' || ch == '\u000B' || ch == '\u000C' || ch == '\r' || ch == ' ' ||
                ch == ' ' || ch == ' ' || ch in ' '..' ' || ch == ' ' ||
                ch == ' ' || ch == ' ' || ch == ' ' || ch == '　' || ch == '﻿'
        }

        fun isJsWhitespace(ch: Char): Boolean = JS_WS_CHARS(ch)
    }
}

/** JS `String.prototype.trim()`. */
val String.jsTrimmed: String get() = trim(JSRegex::isJsWhitespace)

private val WHITESPACE_RUN = JSRegex("\\s+")

/** JS `s.replace(/\s+/g, " ")`. */
val String.collapsedWhitespace: String get() = WHITESPACE_RUN.replaceAll(this, " ")
